File: Data Structures & Algorithms/insert-new-interval/submission-0.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = []
        i = 0
        n = len(intervals)

        # Phase 1: Add all intervals completely before newInterval
        while i < n and intervals[i][1] < newInterval[0]:
            ans.append(intervals[i])
            i += 1

        # Phase 2: Merge all overlapping intervals
        while i < n and intervals[i][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i += 1

        ans.append(newInterval)

        # Phase 3: Add the remaining intervals
        while i < n:
            ans.append(intervals[i])
            i += 1

        return ans

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int i=0, j=0, n=s.size(), ans=0;
        map<char, int>m;
        while(j<n){
            m[s[j]]++;
            int winLen = j-i+1;
            if(m.size()==winLen){
                ans=max(ans, winLen);
            } else {
                m[s[i]]--;
                if(m[s[i]]==0) m.erase(s[i]);
                i++;
            }
            j++;
        }
        return ans;
    }
};
